// @expo/cli/build/src/start/server/metro/debugging/attachAtlas.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "attachAtlasAsync", {
    enumerable: true,
    get: ()=>attachAtlasAsync
});
const _atlasPrerequisite = require("./AtlasPrerequisite");
const _env = require("../../../../utils/env");
const debug = require("debug")("expo:metro:debugging:attachAtlas");
async function attachAtlasAsync(options) {
    if (!_env.env.EXPO_UNSTABLE_ATLAS) {
        return;
    }
    debug("Atlas is enabled, initializing for this project...");
    await new _atlasPrerequisite.AtlasPrerequisite(options.projectRoot).bootstrapAsync({
        exp: options.exp
    });
    return !options.isExporting ? attachAtlasToDevServer(options) : await attachAtlasToExport(options);
}
/**
 * Attach Atlas to the Metro bundler for development mode.
 * This includes attaching to Metro's middleware stack to host the Atlas UI.
 */ function attachAtlasToDevServer(options) {
    if (!options.middleware) {
        throw new Error("Expected middleware to be provided for Atlas when running in development mode");
    }
    const atlas = importAtlasForDev(options.projectRoot);
    if (!atlas) {
        return debug("Atlas is not installed in the project, skipping initialization");
    }
    const instance = atlas.createExpoAtlasMiddleware(options.metroConfig);
    options.middleware.use("/_expo/atlas", instance.middleware);
    debug("Attached Atlas middleware for development on: /_expo/atlas");
    return instance;
}
function importAtlasForDev(projectRoot) {
    try {
        return require(require.resolve("expo-atlas/cli", {
            paths: [
                projectRoot
            ]
        }));
    } catch (error) {
        debug("Failed to load Atlas from project:", error);
        return null;
    }
}
/**
 * Attach Atlas to the Metro bundler for exporting mode.
 * This only includes attaching the custom serializer to the Metro config.
 */ async function attachAtlasToExport(options) {
    const atlas = importAtlasForExport(options.projectRoot);
    if (!atlas) {
        return debug("Atlas is not installed in the project, skipping initialization");
    }
    if (options.resetAtlasFile) {
        const filePath = await atlas.resetExpoAtlasFile(options.projectRoot);
        debug("(Re)created Atlas file at:", filePath);
    }
    atlas.withExpoAtlas(options.metroConfig);
    debug("Attached Atlas to Metro config for exporting");
}
function importAtlasForExport(projectRoot) {
    try {
        return require(require.resolve("expo-atlas/metro", {
            paths: [
                projectRoot
            ]
        }));
    } catch (error) {
        debug("Failed to load Atlas from project:", error);
        return null;
    }
}
